main.go

```diff
@@ -0,0 +1,193 @@
+	}
+
+	if m.Engines.Npm == "" {
+		return "", nil
+	}
+
+	v, err := semver.NewVersion(m.Engines.Npm)
+	if err != nil {
+		return "", fmt.Errorf("engines.npm is not valid semver string")
+	}
+
+	return v.String(), nil
+}
+
+func createInstallNpmCommand(os string) *command.Model {
+	var args []string
+	switch os {
+	case "darwin":
+		args = []string{"brew", "install", "node"}
+	case "linux":
+		args = []string{"apt-get", "-y", "install", "npm"}
+	}
+
+	return command.New(args[0], args[1:]...)
+}
+
+func installLatestNpm() (string, error) {
+	cmd := createInstallNpmCommand(runtime.GOOS)
+
+	var out string
+	out, err := command.RunCmdAndReturnTrimmedOutput(cmd.GetCmd())
+
+	if err != nil {
+		return out, fmt.Errorf("error running npm install: %s", err)
+	}
+
+	return out, nil
+}
+
+func runAndLog(cmd *command.Model) (string, error) {
+	out, err := cmd.RunAndReturnTrimmedCombinedOutput()
+	if err != nil {
+		return out, fmt.Errorf("error running npm command: %s", err)
+	}
+
+	log.Donef(fmt.Sprintf("$ %s", cmd.PrintableCommandArgs()))
+	log.Printf(out)
+
+	return out, nil
+}
+
+func setNpmVersion(ver string) (string, error) {
+	cmd := command.New("npm", "install", "-g", fmt.Sprintf("npm@%s", ver))
+	out, err := runAndLog(cmd)
+	if err != nil {
+		return out, fmt.Errorf("error running npm install: %s", err)
+	}
+
+	return out, nil
+}
+
+func failf(f string, args ...interface{}) {
+	log.Errorf(f, args...)
+	fmt.Println()
+	os.Exit(1)
+}
+
+func main() {
+	var config Config
+	if err := stepconf.Parse(&config); err != nil {
+		failf("error parsing step config: %s", err)
+	}
+	stepconf.Print(config)
+
+	if config.NpmVersion == "" {
+		fmt.Println()
+		log.Infof("Autodetecting npm version")
+
+		log.Printf("Checking package.json for npm version")
+		ver, err := getNpmVersionFromPackageJSON()
+		if err != nil {
+			failf("error processing package.json: %s", err)
+		}
+
+		if ver == "" {
+			log.Warnf("No npm version found in package.json")
+
+			log.Printf("Locating system installed npm")
+			path, err := exec.LookPath("npm")
+			if err != nil {
+				log.Warnf("npm not found on PATH")
+
+				log.Printf("Installing latest npm")
+				ver = "latest"
+
+				out, err := runAndLog(createInstallNpmCommand(runtime.GOOS))
+				if err != nil {
+					log.Errorf(out)
+					failf("Error installing npm: %s", err)
+				}
+
+				log.Printf(out)
+			} else {
+				log.Printf("npm found at %s", path)
+
+				cmd := command.New("npm", "--version")
+				out, err := runAndLog(cmd)
+				if err != nil {
+					log.Warnf("Error getting installed npm version: %s", err)
+				}
+				ver = out
+			}
+		} else {
+			fmt.Println(ver)
+			fmt.Println()
+			log.Infof("Setting npm version to %s", ver)
+
+			out, err := setNpmVersion(ver)
+			if err != nil {
+				log.Errorf(out)
+				failf("Error setting npm version to %s: %s", ver, err)
+			}
+		}
+	} else {
+		fmt.Println()
+		log.Infof("Setting npm version to %s", config.NpmVersion)
+
+		out, err := setNpmVersion(config.NpmVersion)
```

Comment:
since `setNpmVersion ` calls `runAndLog`, which logs the command output, in case of error the output will be logged two times